lms/djangoapps/courseware/courses.py

```diff
@@ -449,10 +449,44 @@ def get_course_syllabus_section(course, section_key):
     raise KeyError("Invalid about key " + str(section_key))
 
 
+class AccessLimitedCourseList(object):
+    def __init__(self, course_queryset, user, permission_name):
+        self.course_queryset = course_queryset
+        self._access_limited_courses = (c for c in self.course_queryset if has_access(user, permission_name, c))
+        self._count = None
+        self._course_cache = []
+
+    def __len__(self):
+        if self._count is None:
+            self._count = self.course_queryset.count()
+        return self._count
+
+    def __iter__(self):
+        for course in self._course_cache:
+            yield course
+        for course in self._access_limited_courses:
+            self._course_cache.append(course)
+            yield course
+
+    def __getitem__(self, index):
+        if isinstance(index, int):
+            while len(self._course_cache) < index:
+                self._course_cache.append(next(self._access_limited_courses))
+            return self._course_cache[index]
+        elif isinstance(index, slice):
+            if index.stop is not None:
+                while len(self._course_cache) < index.stop:
+                    self._course_cache.append(next(self._access_limited_courses))
+            else:
+                self._course_cache.extend(self._access_limited_courses)
+            return self._course_cache[index]
+        else:
+            raise TypeError()
+
+
 def get_courses(user, org=None, filter_=None):
     """
-    Returns a list of courses available, sorted by course.number and optionally
-    filtered by org code (case-insensitive).
+    Yield courses available, optionally filtered by org code (case-insensitive).
```

Comment:
You went to a lot of trouble to change "return" to "yield" in docstrings, which is admirable, but I'm not sure yield is correct here.  We don't have a good way to say, "this will give you an iterable thing, don't worry about the details."